dit/multivariate/base_markov_optimizer.py

```diff
@@ -0,0 +1,582 @@
+
+        Parameters
+        ----------
+        x : ndarray
+            An optimization vector.
+        """
+        joint = self.construct_joint(x)
+
+        # p(rvs)
+        others = joint.sum(axis=-1, keepdims=True)
+
+        # p(w)
+        axes = tuple(range(len(joint.shape)-1))
+        markov_var = joint.sum(axis=axes, keepdims=True)
+
+        mut_info = np.nansum(joint * np.log2(joint / (markov_var * others)))
+
+        return mut_info
+
+    def optimize(self, x0=None, nhops=5, jacobian=False, polish=True):
+        """
+        Perform the optimization.
+
+        Parameters
+        ----------
+        x0 : ndarray, None
+            The vector to initialize the optimization with. If None, a random
+            vector is used.
+        nhops : int
+            The number of times to basin hop in the optimization.
+        jacobian : bool
+            Whether to use an Jacobians computed with numdifftools or not.
+            Defaults to False. The use of numdifftools can greatly slow down the
+            speed of the algorithms.
+        polish : bool
+            Whether to polish the result or not. If True, this will perform a
+            second optimization seeded with the result of the first, but with
+            smaller tolerances.
+
+        Notes
+        -----
+        """
+        from scipy.optimize import basinhopping
+
+        if x0 is not None:
+            x = x0
+        else:
+            x = self.construct_random_initial()
+
+        def accept_test(**kwargs):
+            x = kwargs['x_new']
+            tmax = bool(np.all(x <= 1))
+            tmin = bool(np.all(x >= 0))
+            return tmin and tmax
+
+        minimizer_kwargs = {'method': 'SLSQP',
+                            'bounds': [(0, 1)]*x.size,
+                            'constraints': self.constraints,
+                            'tol': None,
+                            'callback': None,
+                            'options': {'maxiter': 1000,
+                                        'ftol': 5e-07,
+                                        'eps': 1.4901161193847656e-09,
+                                       },
+                           }
+
+        # this makes things slower!
+        # TODO: use ndt.nd_algopy, which may be significantly faster.
+        if jacobian: # pragma: no cover
+            import numdifftools as ndt
+
+            minimizer_kwargs['jac'] = ndt.Jacobian(self.objective)
+            for const in minimizer_kwargs['constraints']:
+                const['jac'] = ndt.Jacobian(const['fun'])
+
+        res = basinhopping(func=self.objective,
+                           x0=x,
+                           minimizer_kwargs=minimizer_kwargs,
+                           niter=nhops,
+                           accept_test=accept_test,
+                          )
+
+        self._res = res
+
+        success, msg = self._success(res)
+        if not success:
+            raise Exception(msg)
+
+        if polish:
```

Comment:
Maybe allow polish to specify the cutoff when it is not False? Or add a cutoff parameter?